sql/src/main/java/io/crate/execution/engine/collect/PKLookupOperation.java

```diff
@@ -145,19 +154,24 @@ public void runWithShardProjections(UUID jobId,
                 ramAccountingContext,
                 shardAndIds.projectorFactory
             );
-            BatchIterator<Row> batchIterator = InMemoryBatchIterator.of(
-                shardAndIds.value.stream()
-                    .map(pkAndVersion -> shardAndIds.shard.getService().get(
-                        Constants.DEFAULT_MAPPING_TYPE,
-                        pkAndVersion.id(),
-                        emptyFields,
-                        true,
-                        pkAndVersion.version(),
-                        VersionType.EXTERNAL,
-                        FetchSourceContext.FETCH_SOURCE
-                    ))
-                    .map(resultToRow)
-                    ::iterator, null);
+            Stream<Row> rowStream = shardAndIds.value.stream()
+                .map(pkAndVersion -> shardAndIds.shard.getService().get(
+                    Constants.DEFAULT_MAPPING_TYPE,
+                    pkAndVersion.id(),
+                    emptyFields,
+                    true,
+                    pkAndVersion.version(),
+                    VersionType.EXTERNAL,
+                    FetchSourceContext.FETCH_SOURCE
+                ))
+                .map(resultToRow);
+            final Iterable<Row> rowIterable;
+            if (consumer.requiresScroll()) {
+                rowIterable = rowStream.collect(Collectors.toList());
```

Comment:
The `resultToRow` function uses a single row instance which is mutated on each call, so collecting the same instance in a list is not a safe operation. Would have to go through another `.map(r -> new RowN(r.materialize())`

Reply:
Ouch, thanks for spotting that.